### ISSUE TITLE
Canvas fails to render when camera attached

### ISSUE THREAD

**Issue by rdlaitila** [vrld/hump#69]
Hello,

I'm running into an issue where within a block of camera attach/detach calls, a canvas fails to render correctly. I'm hoping you can help indicate the issue with my test code or verify that this is a bug with  the expected behavior of hump.camera:

```lua
local attachcam = true -- set to false to see expected behavior
local camera = require('hump.camera')(0,0)
local canvasWithCam = love.graphics.newCanvas(50, 50)
local canvasWithoutCam = love.graphics.newCanvas(50, 50)

function love.draw()        
    -- Works as expected
    love.graphics.setCanvas(canvasWithoutCam)    
    love.graphics.rectangle('fill', 0, 0, 50, 50)    
    love.graphics.setCanvas()  
    love.graphics.draw(canvasWithoutCam, 0, 0)
    
    -- if cam is attached, canvas fails to render
    if attachcam then camera:attach() end
    love.graphics.setCanvas(canvasWithCam)    
    love.graphics.rectangle('fill', 0, 0, 50, 50)    
    love.graphics.setCanvas()  
    love.graphics.draw(canvasWithCam, 100, 0)
    if attachcam then camera:detach() end
end
```

I do not have any viable workaround at this time, but if I develop one I will post my updated code. Thanks for your time! Please let me know if you need any further details.

**Comment by rdlaitila** [comment]
Screenshots of expected/actual:

Expected:

![hump_bug_expected](https://cloud.githubusercontent.com/assets/1124388/16326811/dc95e22e-3997-11e6-8c7a-b95f7195dcfe.png)

Actual:

![hump_bug_actual](https://cloud.githubusercontent.com/assets/1124388/16326814/e37f99b8-3997-11e6-85df-61579364a46e.png)

**Comment by buckle2000** [comment]
@palmettos whether or not scissor is set, opengl takes almost the same time to render.

**Comment by buckle2000** [comment]
Solved.

**Comment by palmettos** [comment]
@buckle2000 Your pull request indicates that it does solve the problem. Thanks for the info on glScissor, though.

**Comment by vrld** [comment]
Well... 

```lua
camera:attach()
love.graphics.setCanvas(canvasWithCam)    
love.graphics.rectangle('fill', 0, 0, 50, 50)    
love.graphics.setCanvas()  
love.graphics.draw(canvasWithCam, 100, 0)
camera:detach()
```
... draws a rectangle to the canvas while the camera transformation is active and also draws the canvas itself, while the camera is active. When not given any other bounding rectangle, `camera:attach()` will assume the viewing window is located at `0,0`, with the dimension `W,H` of the main window. Then the scene is transformed so that the point `0,0` (in camera coordinates) is at position `W/2,H/2` (in screen coordinates). This means the rectangle is actually drawn at position `W/2,H/2`. The canvas is drawn at `W/2+100,H/2`.

But the rectangle is still not visible. Normally, you would expect it to be at position `W/2+100,H/2` (upper left corner of the canvas + camera transformation). However, in your example, the canvas is only 50 x 50 px in size, so the rectangle *is drawn outside the canvas*. If you create the canvas with larger dimension (`local canvasWithCam = love.graphics.newCanvas()`), then the rectangle will be visible.


Now, if you provide the viewing rectangle to `camera:attach()`, it will work as expected (see attached pic):

```lua
camera:attach(0,0,50,50) -- center at 25,25, 
  love.graphics.setCanvas(canvasWithCam)
   -- this will show only a quarter of the rectangle, the rest is outside the canvas
  love.graphics.rectangle('fill', 0, 0, 50, 50)
   -- this will show as expected
  love.graphics.rectangle('line', -25, -25, 50, 50)
  love.graphics.setCanvas()
camera:detach()

love.graphics.draw(canvasWithCam, 100, 0)
```

In short:

- Draw canvas outside of camera transformation
- Use viewing window paramters to `attach()`; match parameters to canvas size
- Do not draw outside of the canvas

**Comment by vrld** [comment]
![cam](https://cloud.githubusercontent.com/assets/156469/18034909/18030e1e-6d4a-11e6-8f8c-7c5c14ddd500.png)

**Comment by buckle2000** [comment]
I believe this is a bug.
https://github.com/vrld/hump/pull/70/commits/fcb0e87d4e4231e34571ed152eed57e5b9c8a662

**Comment by vrld** [comment]
Why do you think this is a bug with hump? How does the change fix the bug?

From my point of view, the real issue is that the rectangle is drawn outside of the canvas. This is a bug, yes, but higher up the hierarchy (i.e. in the game). The proposed change works in this case because it puts `(0,0)` (well, actually `cam.x,cam.y`) at the upper left corner. But this breaks with the current behavior, which puts (0,0) in the center of the camera window.

Also, I wouldn't call one month reaction time fast... ;)
